keyboards/mlego/m20/m20.c

```diff
@@ -0,0 +1,205 @@
+const rgblight_segment_t PROGMEM my_rse_layer[]    = RGBLIGHT_LAYER_SEGMENTS({0, RGBLED_NUM, HSV_RED});
+const rgblight_segment_t PROGMEM my_adj_layer[]    = RGBLIGHT_LAYER_SEGMENTS({0, RGBLED_NUM, HSV_GREEN});
+
+const rgblight_segment_t* const PROGMEM my_rgb_layers[] = RGBLIGHT_LAYERS_LIST(my_qwerty_layer, my_lwr_layer, my_rse_layer, my_adj_layer);
+
+#endif
+
+#ifdef OLED_ENABLE
+
+static uint32_t oled_logo_timer = 0;
+static bool clear_logo = true;
+static const char PROGMEM m65_logo[] = {
+    0x91, 0x92, 0x93, 0x94, 0x95, 0x96, 0x97, 0x98, 0x99, 0xa0, 0xa1, 0xa2, 0xa3, 0xa4, 0xa5, 0xa5, 0xa5, 0xa5, 0xa5, 0xa5, 0xa5,
+    0x80, 0x81, 0x82, 0x83, 0x84, 0x85, 0x86, 0x87, 0x88, 0x89, 0x8a, 0x8b, 0x8c, 0x8d, 0x8e, 0x8f, 0x90, 0x91, 0x92, 0x93, 0x94,
+    0xa0, 0xa1, 0xa2, 0xa3, 0xa4, 0xa5, 0xa6, 0xa7, 0xa8, 0xa9, 0xaa, 0xab, 0xac, 0xad, 0xae, 0xaf, 0xb0, 0xb1, 0xb2, 0xb3, 0xb4,
+    0xc0, 0xc1, 0xc2, 0xc3, 0xc4, 0xc5, 0xc6, 0xc7, 0xc8, 0xc9, 0xca, 0xcb, 0xcc, 0xcd, 0xce, 0xcf, 0xd0, 0xd1, 0xd2, 0xd3, 0xd4,
+    0};
+
+#endif
+
+#ifdef RGBLIGHT_ENABLE
+
+void set_rgb_layers(layer_state_t state){
+
+    rgblight_set_layer_state(0, layer_state_cmp(state, _NUM));
+    rgblight_set_layer_state(1, layer_state_cmp(state, _LWR));
+    rgblight_set_layer_state(2, layer_state_cmp(state, _RSE));
+    rgblight_set_layer_state(3, layer_state_cmp(state, _ADJ));
+
+}
+
+void set_default_rgb_layers(layer_state_t state){
+    rgblight_set_layer_state(0, layer_state_cmp(state, _NUM));
+}
+
+const rgblight_segment_t * const* my_rgb(void){
+    return my_rgb_layers;
+}
+
+#endif
+
+void set_led_toggle(const uint8_t layer, const bool state){
+
+    switch (layer) {
+        case _LWR:
+          toggle_lwr = state;
+          break;
+        case _RSE:
+          toggle_rse = state;
+          break;
+        default:
+          break;
+    }
+}
+
+void toggle_leds(void){
+
+    led_lwr(toggle_lwr);
+    led_rse(toggle_rse);
+      if (layer_state_is(_ADJ)) {
+        led_lwr(true);
+        led_rse(true);
+    }
+
+}
+
+#ifdef ENCODER_ENABLE
+
+#    define MEDIA_KEY_DELAY 10
+
+void my_encoders(const uint8_t index, const bool clockwise) {
+    if (index == 0) { /* First encoder */
+#ifdef RGBLIGHT_ENABLE
+        if (IS_LAYER_ON(_LWR)) {
+            if (clockwise) {
+                rgblight_decrease_val_noeeprom();
+            } else {
+                rgblight_increase_val_noeeprom();
+            }
+        } else if (IS_LAYER_ON(_RSE)) {
+            if (clockwise) {
+                rgblight_decrease_hue_noeeprom();
+            } else {
+                rgblight_increase_hue_noeeprom();
+            }
+
+        } else {
+#endif
+            if (clockwise) {
+                tap_code_delay(KC_VOLD, MEDIA_KEY_DELAY);
+            } else {
+                tap_code_delay(KC_VOLU, MEDIA_KEY_DELAY);
+            }
+#ifdef RGBLIGHT_ENABLE
+        }
+#endif
+    }
+}
```

Comment:
If anything, this should simply be replaced with encoder map support, at the keymap level.

Reply:
no idea about the option, any reference?

Reply:
It's new as of this last breaking changes cycle (0.17.0).  It uses an array like the keymap, but for the encoders, and is handled by "action_exec", so it is capable of using the full process_record stack, including custom keycodes (such as rgb keycodes, etc).

https://docs.qmk.fm/#/feature_encoders?id=encoder-map

There are a number of keymaps that are already using it, such as mine.  But the documents have examples for them. 

That said, it should be noted that it should only be enabled at the keymap level, and the configurator and via don't support it currently, but the underlying "dynamic keymaps" feature that via relies on does support it.

Reply:
thanks, used now and pushed.